### PR TITLE
zmarkdown/rebber-plugins/mdast-split-by-heading: uses introduction commands in latex

### PR DESCRIPTION
**Pull request by artragis**
Here is my PR to implement custom introduction/conclusion commands inside latex document.

This is a WIP because I have some difficulties to run the tests.

Moreover I don't know if useing `\levelOneIntroduction` is good as introductions may have blocks, headings... @karnaj might have a say about it.

### CODE REVIEW

**Comment by TitiAlone** [comment]
`splitted` is oftentimes empty, you should handle this case in `introductionAndConclusionVisitor.js`:

```diff
     const splitted = splitAtDepth(rootNode, currentDepth + 1)
+    if(!splitted || !splitted.subTrees) return (node) => {}
     const conclusionCandidate = splitted.subTrees[splitted.subTrees.length]
```

It should allow the tests to pass, but anyway I'm not convinced with your code right now, will have to take a closer look.

File: packages/rebber-plugins/src/type/introduction.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module.exports = introduction
+const all = require('rebber/dist/all')
+const introductionCommands = ['levelOneIntroduction', 'levelTwoIntroduction',
```
Comment (Karnaj) [inline packages/rebber-plugins/src/type/introduction.js]: `Level`, not `level`.

**Comment by artragis** [comment]
your fix does not change anything.

**Comment by TitiAlone** [comment]
Solves all failing tests on my local checkout, maybe you forgot to build? What's for sure is that you didn't push the changes.

**Comment by artragis** [comment]
Ok, the problem was I did not returned (node) => {} even if I don't understand such a return

File: packages/rebber/dist/escaper.js [hunk 2]
```diff
@@ -22,9 +22,8 @@ var defaultEscapes = {
   '{': '\\{',
   '}': '\\}',
   '~': '\\textasciitilde{}'
-  /* Encode special characters in `value`. */
-
 };
+/* Encode special characters in `value`. */
```
Comment (TitiAlone) [inline packages/rebber/dist/escaper.js]: No changes here, please

File: packages/remark-custom-blocks/dist/index.js [hunk 3]
```diff
@@ -122,11 +122,11 @@ module.exports = function blockPlugin() {
     var blockChildren = [contents];
 
     if (titleAllowed && blockTitle) {
-      var configuredTitleBlockType = potentialBlock.titleElement ? potentialBlock.titleElement : 'div';
+      var titleElement = potentialBlock.titleElement ? potentialBlock.titleElement : 'div';
```
Comment (TitiAlone) [inline packages/remark-custom-blocks/dist/index.js]: Why is this here? Wasn't it merged with the other PR?

File: packages/zmarkdown/server/handlers.js [hunk 4]
```diff
@@ -162,6 +164,14 @@ module.exports = function markdownHandlers (Raven) {
         ])
       }
 
+      if (opts.content_type === 'big') {
+        rebber.preprocessors.root = introductionAndConclusionVisitor(2)
+      } else if (opts.content_type === 'middle') {
+        rebber.preprocessors.root = introductionAndConclusionVisitor(1)
+      } else {
+        rebber.preprocessors.root = introductionAndConclusionVisitor(0)
+      }
```
Comment (TitiAlone) [inline packages/zmarkdown/server/handlers.js]: I feel like having named arguments would be way easier to understand.

File: packages/rebber-plugins/src/preprocessors/introductionAndConclusionVisitor.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+const splitAtDepth = require('mdast-util-split-by-heading/dist')
+
+function introductionAndConclusionVisitor (headingDepth = 0, currentDepth = 0) {
+  return (ctx, rootNode) => {
+    const splitted = splitAtDepth(rootNode, currentDepth + 1)
+    const conclusionCandidate = splitted.subTrees[splitted.subTrees.length]
```
[the remark this answers is not in the document]
Comment (artragis) [reply packages/rebber-plugins/src/preprocessors/introductionAndConclusionVisitor.js]: Nice catch !

File: packages/rebber/dist/escaper.js (hunk 2 above)
Comment (artragis) [reply packages/rebber/dist/escaper.js]: grr this is learna build that creates that.

Comment (TitiAlone) [reply packages/rebber/dist/escaper.js]: You can run it locally using `--scope=rebber-plugins`

Comment (artragis) [reply packages/rebber/dist/escaper.js]: As I needed many plugins, I did not use scope but this is really annoying if the comman dresult is not consistant accross computers.
